token/js/test/unit/tlvData.test.ts

```diff
@@ -0,0 +1,117 @@
+        Buffer.from([2, 0]),
+        Buffer.from([1, 2]),
+    ]);
+
+    // typeLength 2, lengthSize 1
+    const tlvData2 = Buffer.concat([
+        Buffer.from([0, 0]),
+        Buffer.from([0]),
+        Buffer.from([]),
+        Buffer.from([1, 0]),
+        Buffer.from([1]),
+        Buffer.from([1]),
+        Buffer.from([2, 0]),
+        Buffer.from([2]),
+        Buffer.from([1, 2]),
+    ]);
+
+    // typeLength 4, lengthSize 8
+    const tlvData3 = Buffer.concat([
+        Buffer.from([0, 0, 0, 0]),
+        Buffer.from([0, 0, 0, 0, 0, 0, 0, 0]),
+        Buffer.from([]),
+        Buffer.from([1, 0, 0, 0]),
+        Buffer.from([1, 0, 0, 0, 0, 0, 0, 0]),
+        Buffer.from([1]),
+        Buffer.from([2, 0, 0, 0]),
+        Buffer.from([2, 0, 0, 0, 0, 0, 0, 0]),
+        Buffer.from([1, 2]),
+    ]);
+
+    // typeLength 8, lengthSize 4
+    const tlvData4 = Buffer.concat([
+        Buffer.from([0, 0, 0, 0, 0, 0, 0, 0]),
+        Buffer.from([0, 0, 0, 0]),
+        Buffer.from([]),
+        Buffer.from([1, 0, 0, 0, 0, 0, 0, 0]),
+        Buffer.from([1, 0, 0, 0]),
+        Buffer.from([1]),
+        Buffer.from([2, 0, 0, 0, 0, 0, 0, 0]),
+        Buffer.from([2, 0, 0, 0]),
+        Buffer.from([1, 2]),
+    ]);
+
+    const testRawData = (tlvData: Buffer, typeSize: TLVNumberSize, lengthSize: TLVNumberSize) => {
+        const tlv = new TLVData(tlvData, typeSize, lengthSize);
+        expect(tlv.data).to.be.deep.equal(tlvData);
+        const tlvWithOffset = new TLVData(tlvData, typeSize, lengthSize, typeSize + lengthSize);
+        expect(tlvWithOffset.data).to.be.deep.equal(tlvData.subarray(typeSize + lengthSize));
+    };
+
+    it('should get the raw tlv data', () => {
+        testRawData(tlvData1, 1, 2);
+        testRawData(tlvData2, 2, 1);
+        testRawData(tlvData3, 4, 8);
+        testRawData(tlvData4, 8, 4);
+    });
+
+    const testIndividualEntries = (tlvData: Buffer, typeSize: TLVNumberSize, lengthSize: TLVNumberSize) => {
+        const tlv = new TLVData(tlvData, typeSize, lengthSize);
+        expect(tlv.entry(Number(0))).to.be.deep.equal(Buffer.alloc(0));
+        expect(tlv.entry(Number(1))).to.be.deep.equal(Buffer.from([1]));
+        expect(tlv.entry(Number(2))).to.be.deep.equal(Buffer.from([1, 2]));
+        expect(tlv.entry(Number(3))).to.be.null;
+        expect(tlv.entry(BigInt(0))).to.be.deep.equal(Buffer.alloc(0));
+        expect(tlv.entry(BigInt(1))).to.be.deep.equal(Buffer.from([1]));
+        expect(tlv.entry(BigInt(2))).to.be.deep.equal(Buffer.from([1, 2]));
+        expect(tlv.entry(BigInt(3))).to.be.null;
+
+        const type = Buffer.alloc(typeSize);
+        type[0] = 0;
+        expect(tlv.entry(type)).to.be.deep.equal(Buffer.alloc(0));
+        type[0] = 1;
+        expect(tlv.entry(type)).to.be.deep.equal(Buffer.from([1]));
+        type[0] = 2;
+        expect(tlv.entry(type)).to.be.deep.equal(Buffer.from([1, 2]));
+        type[0] = 3;
+        expect(tlv.entry(type)).to.be.null;
+    };
```

Comment:
Am I reading this correctly in that it appears you've combined grabbing by index or by discriminator?

That's pretty interesting. I guess someone would have to know the ordering of the TLV entries, but if they do it's not bad to have. I would sort of lean toward a separate function though, for brevity. Curious your thoughts.

With that being said I wonder if there's a similarly slick way to do grabbing an entry for a certain discriminator (that repeats) at a certain repetition number? That's what this bad boy does. Given a repetition number, say 2, it will return the second TLV entry for that discriminator:

https://github.com/solana-labs/solana-program-library/blob/f35dc5fdb6fc01d054afc51a6ea5a10307a48e3b/libraries/type-length-value/src/state.rs#L243-L249

Reply:
The `entry` method currently only works by discriminator (not by index). It transforms anything (number, bigint, bytes, etc.) you provide into raw bytes and then matches that to each discriminator it encounters. It might be nice to add grabbing by index but I think that for now this is not (yet) needed?

Reply:
The Rust library supports either grabbing the first or grabbing by which ever "repetition count" (for repeating entries), but it's currently not used by anything in SPL.

For that reason I can live without it for now, but it should eventually go in. Doesn't have to be this PR though